text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+    };
+  }
+}
+```
+
+This allows the manager to store metadata that it doesn't want to expose to the
+user.
+
+This hook has the following timing semantics:
+
+**Always**
+- called as discovered during DOM construction
+- called in definition order in the template
+
+#### `getValue`
+
+`getValue` is an optional hook that should return the value of the helper. This
+is the value that is returned from the helper and passed into the template.
+
+This hook is called when the value is requested from the helper (e.g. when the
+template is rendering and the helper value is needed). The hook is autotracked,
+and will rerun whenever any tracked values used inside of it are updated.
+Otherwise it does not rerun.
+
+> Note: This means that arguments which are not _consumed_ within the hook will
+> not trigger updates.
+
+This hook is only called for helpers with the `hasValue` capability enabled.
+This hook has the following timing semantics:
+
+**Always**
+- called the first time the helper value is requested
+- called after autotracked state has changed
+
+**Never**
+- called if the `hasValue` capability is disabled
+
+#### `runEffect`
+
+`runEffect` is an optional hook that should run the effect that the helper is
+applying, setting it up or updating it.
+
+This hook is scheduled to be called after render. The hook is autotracked, and
+will rerun whenever any tracked values used inside of it are updated. Otherwise
+it does not rerun.
+
+The hook is also run during a time period where state mutations are _disabled_
+in Ember. Any change to any tracked property or tag via `Ember.set` will throw
+an error during this time. This is meant to prevent infinite rerenders and other
+antipatterns.
+
+This hook is only called for helpers with the `isScheduledEffect` capability
+enabled. It has the following timing semantics:
+
+**Always**
+- called after the helper was first created
+- called after autotracked state has changed
+
+**Never**
+- called if the `isScheduledEffect` capability is disabled
+
+#### `getDestroyable`
+
+`getDestroyable` is an optional hook that users can use to register a
+destroyable object for the helper. This destroyable will be registered to the
+containing block or template parent, and will be destroyed when it is destroyed.
+See the [Destroyables RFC](https://github.com/emberjs/rfcs/blob/master/text/0580-destroyables.md)
+for more details.
+
+`getDestroyable` is only called if the `hasDestroyable` capability is enabled.
+
+This hook has the following timing semantics:
+
+**Always**
+- called immediately after the `createHelper` hook is called
+
+**Never**
+- called if the `hasDestroyable` capability is disabled
+
+### Capabilities
+
+There are three proposed capabilities for helper managers:
+
+* `hasDestroyable`
+* `isScheduledEffect`
+* `hasValue`
+
+Out of these capabilities, one of `isScheduledEffect` or `hasValue` _must_ be
+enabled. The other must _not_ be enabled, meaning they are mutually exclusive.
```

Comment:
Related, perhaps we can explicitly specify the type (and presumably make sure we ship types for it one way or another—via Typed Ember support or via the package itself, at Framework Core’s discretion) as a union type?

Reply:
FWIW, for effects, using untracked state is actually something that may not be unusual. The point of an effect is to push state out of the system, into an untracked world. If there is a coalescing intermediate state that is untracked, that's perfectly fine, as long as the rest of the app does not need to _react_ to those changes. And if it _does_, then it begs the question of why you are using an effect in the first place.

Reply:
> FWIW, for effects, using untracked state is actually something that may not be unusual.

Yes, I totally understand/agree, but if you go this route it is fundamentally impossible to _also_ allow title tokens to be pushed into the service directly (which is a thing you can do in `ember-page-title` today).

Reply:
> also allow title tokens to be pushed into the service directly (which is a thing you can do in ember-page-title today).

I'm not sure this is true, based on looking at the implementation. `{{title}}` schedules an update from within the helper, and _only_ from within the helper, for changes. That update happens async in the next actions queue, and is the way that `page-title` currently avoids the backtracking rerender assertion (essentially opting into a second render pass).

It'd be possible to do this from a service as well, but I don't think it's possible to avoid a second render pass in general using a declarative approach. By the time you hit a `{{title}}` helper, the actual header has already rendered, so it will necessarily cause a backtracking assertion.

IMO It is _completely fine_ to require a second render pass, and effects as spec'd would definitely allow this. In fact, I don't think there is a way to _disallow_ this. This is the strategy I'll take in the example implementation. The main thing we want is people to do this consciously and carefully, and to document that it is an advanced technique which should be used sparingly (basically, don't just start scheduling willy-nilly if you hit the assertion)

Reply:
I think this is why the demo of ember-page-title will be helpful. In this case, I could easily see the solution being to use two helpers. But what your really want is:

* `{{title` helper just does the pushing of tokens onto the services tracked "title tokens" array
* another helper that actually does the effect that you have to included yourself in `application.hbs` (e.g. `{{set-document-title}}`), this one leverages the tracked title tokens, so it always re-runs the effect as needed)

But you don't _really_ want the first thing to produce a value (the value is meaningless, and the helper ATM returns `''` because it can't signify that it returns nothing). IMHO, it may point to the fact that you want to do "something" during the render cycle, but still not return a value.

Reply:
Sorry, I was basing my last comment on the structure of `{{page-title}}` specifically, which uses `ember-cli-head`, which renders a full template snippet in the header. So mutating a value during render would not be possible, it really does consume the value first, and `{{title}}` really does need to use untracked state and schedule an update itself.

More generally though, I could see the pattern of mutating in a value-helper and then using an effect-helper to finalize the effect. You can compose a value helper and effect helper together to get this property if needed, so I think mutual exclusivity still makes sense.

Reply:
Ya, but in the case that I'm referring to I just don't think producing a value actually makes. One helper is for collecting the tokens, the other is for queuing the side effect, and the one that collects tokens really has no valid "value".

Reply:
Can you make that case more concrete by providing an example helper that I can also convert? Currently I'm focusing on `ember-page-title`, which is not using that strategy, and I think a solid example use case would really help. 

Adding a hook explicitly for side-effects that _does_ allow internal state mutation would essentially be bringing back observers. When we previously discussed this, everyone agreed that would probably not be ideal, and that was why the limitation about mutating state was placed on effect-helpers. We can always remove that restriction in the future, if we find motivating use cases. Users using the `getValue()` hook to frequently circumvent the restriction would, for instance, definitely show that it would have value, as would users frequently scheduling changes from effects directly.

Reply:
Another thing to think about, is that once we add `invokeHelper` into the mix, `getValue()` will _not_ trigger unless used directly in JS. This really implies that side-effects _should not_ happen during render, since there's no way to ensure they happen predictably.

I think we would need to add a _second_ effect queue if we were to move this direction.

1. The first effect queue _would_ allow internal app state changes, and would basically be the same as async observers today.
2. The second queue _would not_ allow internal app state changes, and it would be about reflecting state elsewheres.

Reply:
To be clear, I don't think my concern here is blocking. Just something we need to have a good answer / explanation for. Starting out more restrictive then relaxing that restriction is absolutely better than starting out too permissive and being stuck.

Reply:
Also, either way having the `ember-page-title` case study will help and be a nice justification for the new features.